FSDP parameter sync

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77492
* #77491

Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

rohan-varma committed:
FSDP parameter sync

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] FSDP parameter sync"




[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] FSDP parameter sync"




[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module as well as non-redundant loading of checkpoints from CPU.

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -35,6 +35,17 @@ def _get_full_detached_param(fsdp_model: FullyShardedDataParallel):
+    module_states.extend([buffer.detach().cpu() for buffer in model.buffers()])
+    world_size = dist.get_world_size(process_group)
+    olist = [None for _ in range(world_size)]
+    dist.all_gather_object(olist, module_states, process_group=process_group)
```
```suggestion
    dist.all_gather_object(olist, module_states, group=process_group)
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -567,6 +578,15 @@ class FullyShardedDataParallel(nn.Module):
+        sync_module_states (bool): If ``True``, each individually wrapped FSDP unit will broadcast
+            module parameters from rank 0 to ensure they are the same across all ranks after
+            initialization. This helps ensure model parameters are the same across ranks
+            before starting training, but adds communication overhead to ``__init__``, as atleast
```
```suggestion
            before starting training, but adds communication overhead to ``__init__``, as at least
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -751,6 +773,24 @@ def _run_param_init_fn():
+
+            _sync_params_and_buffers(
+                process_group=self.process_group,
+                # TODO: FSDP should sync buffers as well
```
Is this TODO fixed now since it seems that buffers are being synced?

rohan-varma replied:
Yep, should be fixed

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -751,6 +773,24 @@ def _run_param_init_fn():
+                # TODO: FSDP should sync buffers as well
+                module_states=[param.detach() for param in params].extend(bufs_to_sync),
+                # Same bucket size as DDP
+                broadcast_bucket_size=int(250 * 1024 * 1024),
```
nit: Shall we factor out this constant into a global variable?

awgu commented on torch/distributed/utils.py:
```diff
@@ -31,7 +34,7 @@ def to_map(obj):
                     # Ensure tensor memory is not reused until work on
                     # main stream is complete
-                    output.record_stream(current_stream)  # type: ignore[arg-type]
+                    output.record_stream(current_stream)  # type: ignore[attr-defined]
```
Was the mypy ignore comment meant to be changed? (asking since now mypy is complaining)

awgu reviewed (approved):
LGTM! Approving to unblock, but just double check that the tests in `test_fsdp_misc.py` do pass after fixing the `all_gather_object()` call.

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
looks good!

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge

rohan-varma committed:
Update on "FSDP parameter sync"


Add a `sync_module_states` flag to FSDP which enables each rank to start off with the same module parameters if the user did not seed the same across ranks.

In addiiton, this PR helps with non-redundant loading of checkpoints from CPU. User can load checkpoint into CPU module before wrapping with FSDP on only rank 0, and then use FSDP to automatically synchronize the module across ranks.

Follow-up: Incur only one communication when auto-wrapping, because we can detect the root and kick off all communication only for the root.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge

NivekT commented:
Rebasing onto this PR is causing the following error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pytorch/torch/__init__.py", line 780, in <module>
    from .functional import *  # noqa: F403
  File "/Users/pytorch/torch/functional.py", line 7, in <module>
    import torch.nn.functional as F
  File "/Users/pytorch/torch/nn/__init__.py", line 3, in <module>
    from .parallel import DataParallel
  File "/Users/pytorch/torch/nn/parallel/__init__.py", line 5, in <module>
    from .distributed import DistributedDataParallel
  File "/Users/pytorch/torch/nn/parallel/distributed.py", line 21, in <module>
    from torch.distributed.utils import (
  File "/Users/pytorch/torch/distributed/utils.py", line 121, in <module>
    process_group: dist.ProcessGroup,
AttributeError: module 'torch.distributed' has no attribute 'ProcessGroup'
```
Is this happening to anyone else?

NivekT commented:
Update: talked to @rohan-varma offline. The issue arises when `USE_DISTRIBUTED=0` is used during the build process. One solution may be to check `if dist.is_available()` before importing the relevant attributes.
